### PR TITLE
Add support for RFC 8970 — Message Preview Generation

### PR DESCRIPTION
**Pull request by maxoly**
Add support for IMAP Message Preview Generation — [RFC 8970](https://datatracker.ietf.org/doc/html/rfc8970)

### Motivation:

The current IMAP implementation lacks support for the Message Preview Generation extension as defined in RFC 8970. This extension allows clients to request and receive preview text for email messages directly from the server, improving the user experience by enabling quicker message overviews without the need to download full messages. Implementing this feature addresses the need for efficient email browsing and management, particularly beneficial for applications with bandwidth or storage constraints.

### Modifications:

- Added a new `PREVIEW` capability in the `Capability`.
- Added a new `.preview(lazy: Bool)` fetch attribute
- Added a new `.preview(Preview?)` message attribute
- Updated the `GrammarParser` to handle responses for the `PREVIEW` fetch attribute, extracting the preview text from FETCH responses.
- Added unit tests to validate the functionality of message preview fetching, including command construction and response parsing.

### Result:

After this change, clients using the updated library can request and receive message previews from servers supporting RFC 8970. This enhancement enables more efficient email browsing experiences, allowing users to quickly scan through message summaries without downloading entire messages.

### CODE REVIEW

**Comment by maxoly** [comment]
Hi @danieleggert , Is there any chance that this PR and the others can be evaluated?

**Review by danieleggert — CHANGES_REQUESTED** [review]
Since the RFC states that this `MUST NOT be content transfer encoded and MUST be encoded in UTF-8`, I think this should be decoded into a `String`.

I’d suggest adding a new wrapper type:
```swift
public struct PreviewText: Hashable {
    fileprivate var text: String
    public init(_ text: String) {
        self.text = text
    }
}

extension String {
    public init(_ other: PreviewText) {
        self = other.text
    }
}
```
and then using that throughout the code. That adds (a tiny bit of) type-safety.

File: Sources/NIOIMAPCore/Grammar/Message/MessageAttributes.swift [hunk 1]
```diff
@@ -58,6 +58,12 @@ public enum MessageAttribute: Hashable {
 
     /// `X-GM-LABELS`: provides the labels for a given message
     case gmailLabels([GmailLabel])
+    
+    /// `PREVIEW`: a server-generated abbreviated text representation of message data that
+    /// is useful as a contextual preview of the entire message. 
+    ///
+    /// - SeeAlso: RFC 8970 (IMAP4 Extension: Message Preview Generation)
+    case preview(PreviewText)
```
Comment (danieleggert) [inline Sources/NIOIMAPCore/Grammar/Message/MessageAttributes.swift:66]: If the server responds something like
```
* 1 FETCH (UID 5647 PREVIEW NIL)
```
this should end up being `.preview(nil)` and as such this needs to be `Optional`:
```suggestion
    case preview(PreviewText?)
```

To do this `writeMessageAttribute_preview()` and `parseMessageAttribute_preview()` need to be updated. See below.

File: Tests/NIOIMAPCoreTests/Parser/Grammar/GrammarParser+Fetch+Tests.swift [hunk 2]
```diff
@@ -69,6 +71,7 @@ extension GrammarParser_Fetch_Tests {
                 ("FLAGS (\\seen)", " ", .simpleAttribute(.flags([.seen])), #line),
                 ("FLAGS (\\seen \\answered \\draft)", " ", .simpleAttribute(.flags([.seen, .answered, .draft])), #line),
                 (")\r\n", " ", .finish, #line),
+                ("PREVIEW \"Lorem ipsum dolor sit amet\"", " ", .simpleAttribute(.preview(.init("Lorem ipsum dolor sit amet"))), #line),
```
Comment (danieleggert) [inline Tests/NIOIMAPCoreTests/Parser/Grammar/GrammarParser+Fetch+Tests.swift:74]: This should have a test case for `NIL` and also for the the literal case (if possible).

Here’s an example from the RFC:
```
  S: * 1 FETCH (RFC822.SIZE 5647 PREVIEW {200}
  S: Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  S: Curabitur aliquam turpis et ante dictum, et pulvinar dui congue.
  S: Maecenas hendrerit, lorem non imperdiet pellentesque, nulla
  S: ligula nullam
  S: )
```

**Review by danieleggert — CHANGES_REQUESTED** [review]
Yes, this is great. In re-reading it, I noticed that the `NIL` case isn’t handled, though.

**Review by danieleggert — COMMENTED** [review]
lgtm, but would be good for someone else to take a look.

**Review by Lukasa — APPROVED** [review]
Also looks just fine to me.

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Comment by Lukasa** [comment]
There are some formatting issues that need to be cleaned up:


```diff
diff --git a/Sources/NIOIMAPCore/Grammar/Capability.swift b/Sources/NIOIMAPCore/Grammar/Capability.swift
index eaaf1c5..286a1c5 100644
--- a/Sources/NIOIMAPCore/Grammar/Capability.swift
+++ b/Sources/NIOIMAPCore/Grammar/Capability.swift
@@ -350,7 +350,7 @@ extension Capability {
 
     /// RFC 7888 LITERAL-
     public static let literalMinus = Self(unchecked: "LITERAL-")
-    
+
     /// RFC 8970 - IMAP4 Extension: Message Preview Generation
     ///
     /// Allows a client to request a server-generated abbreviated text representation of message data.
diff --git a/Sources/NIOIMAPCore/Grammar/FetchAttribute.swift b/Sources/NIOIMAPCore/Grammar/FetchAttribute.swift
index d0966b0..5a32279 100644
--- a/Sources/NIOIMAPCore/Grammar/FetchAttribute.swift
+++ b/Sources/NIOIMAPCore/Grammar/FetchAttribute.swift
@@ -68,7 +68,7 @@ public enum FetchAttribute: Hashable {
 
     /// The message's GMail labels.
     case gmailLabels
-    
+
     /// The RFC 8970 Server-generated abbreviated text representation of message
     /// data that is useful as a contextual preview of the entire message.
     case preview(lazy: Bool)
@@ -191,7 +191,7 @@ extension EncodeBuffer {
     @discardableResult mutating func writeFetchAttribute_rfc822Text() -> Int {
         self.writeString("RFC822.TEXT")
     }
-    
+
     @discardableResult mutating func writeFetchAttribute_preview(_ lazy: Bool) -> Int {
         self.writeString(lazy ? "PREVIEW (LAZY)" : "PREVIEW")
     }
diff --git a/Sources/NIOIMAPCore/Grammar/Message/MessageAttributes.swift b/Sources/NIOIMAPCore/Grammar/Message/MessageAttributes.swift
index 4519f59..0824bed 100644
--- a/Sources/NIOIMAPCore/Grammar/Message/MessageAttributes.swift
+++ b/Sources/NIOIMAPCore/Grammar/Message/MessageAttributes.swift
@@ -58,9 +58,9 @@ public enum MessageAttribute: Hashable {
 
     /// `X-GM-LABELS`: provides the labels for a given message
     case gmailLabels([GmailLabel])
-    
+
     /// `PREVIEW`: a server-generated abbreviated text representation of message data that
-    /// is useful as a contextual preview of the entire message. 
+    /// is useful as a contextual preview of the entire message.
     ///
     /// - SeeAlso: RFC 8970 (IMAP4 Extension: Message Preview Generation)
     case preview(PreviewText?)
@@ -203,10 +203,10 @@ extension EncodeBuffer {
                 buffer.writeGmailLabel(label)
             }
     }
-    
+
     @discardableResult mutating func writeMessageAttribute_preview(_ previewText: PreviewText?) -> Int {
         self.writeString("PREVIEW") +
             self.writeSpace() +
-        self.writeNString(previewText.map { String($0) })
+            self.writeNString(previewText.map { String($0) })
     }
 }
diff --git a/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Fetch.swift b/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Fetch.swift
index 6c78d12..1995494 100644
--- a/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Fetch.swift
+++ b/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Fetch.swift
@@ -115,8 +115,8 @@ extension GrammarParser {
             let sectionBinary = try self.parseSectionBinary(buffer: &buffer, tracker: tracker)
             return .binarySize(section: sectionBinary)
         }
-        
-        func parseFetchAttribute_preview(buffer: inout ParseBuffer, tracker: StackTracker) throws -> FetchAttribute {            
+
+        func parseFetchAttribute_preview(buffer: inout ParseBuffer, tracker: StackTracker) throws -> FetchAttribute {
             // RFC 8970
             // fetch-att         =/ "PREVIEW" [SP "(" preview-mod *(SP preview-mod) ")"]
             // preview-mod       =  "LAZY"
@@ -128,7 +128,7 @@ extension GrammarParser {
                     return false
                 }
             }
-            
+
             let lazy = try parsePreviewLazyModifier(buffer: &buffer, tracker: tracker)
             return .preview(lazy: lazy)
         }
@@ -152,7 +152,7 @@ extension GrammarParser {
             "BINARY.SIZE": parseFetchAttribute_binarySize,
             "BINARY": parseFetchAttribute_binary,
             "BINARY.PEEK": parseFetchAttribute_binaryPeek,
-            "PREVIEW": parseFetchAttribute_preview
+            "PREVIEW": parseFetchAttribute_preview,
         ]
 
         // try to use the lookup table, however obviously an unknown number
diff --git a/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Message.swift b/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Message.swift
index f9f5978..937c053 100644
--- a/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Message.swift
+++ b/Sources/NIOIMAPCore/Parser/Grammar/GrammarParser+Message.swift
@@ -206,14 +206,14 @@ extension GrammarParser {
             try self.parseNil(buffer: &buffer, tracker: tracker)
             return .nilBody(kind)
         }
-        
+
         func parseMessageAttribute_preview(buffer: inout ParseBuffer, tracker: StackTracker) throws -> MessageAttribute {
             func parseMessageAttribute_preview_literal(buffer: inout ParseBuffer, tracker: StackTracker) throws -> MessageAttribute {
                 try PL.parseSpaces(buffer: &buffer, tracker: tracker)
                 let raw = try self.parseLiteral(buffer: &buffer, tracker: tracker)
                 return .preview(PreviewText(String(buffer: raw)))
             }
-            
+
             func parseMessageAttribute_preview_inline(buffer: inout ParseBuffer, tracker: StackTracker) throws -> MessageAttribute {
                 try PL.parseSpaces(buffer: &buffer, tracker: tracker)
                 let raw = try self.parseQuoted(buffer: &buffer, tracker: tracker)
@@ -249,7 +249,7 @@ extension GrammarParser {
             "RFC822.TEXT": parseMessageAttribute_rfc822Text_nilBody,
             "RFC822.HEADER": parseMessageAttribute_rfc822Header_nilBody,
             "BINARY": parseMessageAttribute_binary_nilBody,
-            "PREVIEW": parseMessageAttribute_preview
+            "PREVIEW": parseMessageAttribute_preview,
         ]
         return try self.parseFromLookupTable(buffer: &buffer, tracker: tracker, parsers: parsers)
     }
diff --git a/Sources/NIOIMAPCore/PreviewText.swift b/Sources/NIOIMAPCore/PreviewText.swift
index 77957b1..1212d8c 100644
--- a/Sources/NIOIMAPCore/PreviewText.swift
+++ b/Sources/NIOIMAPCore/PreviewText.swift
@@ -14,7 +14,7 @@
 
 public struct PreviewText: Hashable {
     fileprivate let text: String
-    
+
     public init(_ text: String) {
         self.text = text
     }
diff --git a/Tests/NIOIMAPCoreTests/Grammar/FetchAttributeTests.swift b/Tests/NIOIMAPCoreTests/Grammar/FetchAttributeTests.swift
index 0d8b6f6..9008ac5 100644
--- a/Tests/NIOIMAPCoreTests/Grammar/FetchAttributeTests.swift
+++ b/Tests/NIOIMAPCoreTests/Grammar/FetchAttributeTests.swift
@@ -45,7 +45,7 @@ extension FetchAttributeTests {
             (.gmailThreadID, .rfc3501, "X-GM-THRID", #line),
             (.gmailLabels, .rfc3501, "X-GM-LABELS", #line),
             (.preview(lazy: false), .rfc3501, "PREVIEW", #line),
-            (.preview(lazy: true), .rfc3501, "PREVIEW (LAZY)", #line)
+            (.preview(lazy: true), .rfc3501, "PREVIEW (LAZY)", #line),
         ]
         self.iterateInputs(inputs: inputs.map { ($0, $1, [$2], $3) }, encoder: { self.testBuffer.writeFetchAttribute($0) })
     }
@@ -97,7 +97,7 @@ extension FetchAttributeTests {
             ([.flags, .bodyStructure(extensions: false), .rfc822Size, .internalDate, .envelope, .uid], .rfc3501, "(FLAGS BODY RFC822.SIZE INTERNALDATE ENVELOPE UID)", #line),
             ([.gmailLabels, .gmailMessageID, .gmailThreadID], .rfc3501, "(X-GM-LABELS X-GM-MSGID X-GM-THRID)", #line),
             ([.preview(lazy: false)], .rfc3501, "(PREVIEW)", #line),
-            ([.preview(lazy: true)], .rfc3501, "(PREVIEW (LAZY))", #line)
+            ([.preview(lazy: true)], .rfc3501, "(PREVIEW (LAZY))", #line),
         ]
         self.iterateInputs(inputs: inputs.map { ($0, $1, [$2], $3) }, encoder: { self.testBuffer.writeFetchAttributeList($0) })
     }
```

The 5.8 test failure looks to be a flaky test.

**Comment by danieleggert** [comment]
@swift-server-bot test this please

**Review by danieleggert — APPROVED** [review]

**Comment by danieleggert** [comment]
@swift-server-bot test this please

**Pull request merged**